lib/sycamore/sycamore/query/result.py

```diff
@@ -0,0 +1,45 @@
+import io
+from typing import Any, Dict, Optional
+
+from pydantic import BaseModel
+
+from sycamore.query.logical_plan import LogicalPlan
+from sycamore import DocSet
+
+
+class SycamoreQueryResult(BaseModel):
+    """Represents a result from a Sycamore Query operation."""
+
+    query_id: str
+    """The unique ID of the query operation."""
+
+    plan: LogicalPlan
+    """The logical query plan that was executed."""
+
+    result: Any
+    """The result of the query operation. Depending on the query, this could be a list of documents,
+    a single document, a string, an integer, etc.
+    """
+
+    code: Optional[str] = None
+    """The Python code corresponding to the query plan."""
+
+    trace_dirs: Optional[Dict[int, Optional[str]]] = None
```

Comment:
This is essentially a copy of the logical_plan.nodes with runtime information added to it. It's just cache path here but I want to add metrics (at least counts and latency) soon too.

Does it make sense to have a ExecutionResult style object that contains all this? Overall I think we want query level information (basically this object), and then finer-access to individual nodes. 

We can also choose to dump everything in the trace dir (like the [query-plan file](https://github.com/aryn-ai/sycamore/blob/9e2afb8531a31b9f19b4cad762eac3d01fe60ce6/lib/sycamore/sycamore/query/execution/sycamore_executor.py#L272)) but we'd still need an objet that deserializes those files. It would be good for the API response to just give you that stuff

Reply:
If we do that we should probably also rename the variable to execution dir or something

Reply:
Yeah, I thought we might want to just put cache_dir into the LogicalPlan, but them we're mixing logical plan and physical execution stuff. Ideally the plan is "frozen" once generated and we don't manipulate it afterwards, but you're right we need to capture things like metrics and stuff.

I don't have a strong feeling about this. We could make this type an "ExecutionNode" which we could extend with metrics etc, or we could fold it into LogicalPlan. WDYT?

Reply:
Seems like we need a Query object? Where a GetQuery gives you:

```
class Query:
  LogicalPlan
  configs (cache_dir, codegen, llm)
  execution_info (log pointers, node->cached data, metrics)
```

or something along those lines?

Reply:
See what I've done with the NodeExecution type.